# Assignment 1

In [1]:
import os
import numpy as np
import pandas as pd

In [2]:
os.chdir("C:\\recommender_system\\course1\\week2")

In [3]:
mydata = pd.read_csv("HW1-data.csv")

In [4]:
mydata.shape

(20, 22)

In [5]:
mydata.head()

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0   755                   0                                             1.0   
1  5277                   0                                             5.0   
2  1577                   1                                             NaN   
3  4388                   0                                             NaN   
4  1202                   1                                             4.0   

   1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                5.0         
1                                                3.0         
2                                                NaN         
3                                                3.0         
4                                                3.0         

   356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                       2.0                                    NaN   
1                       NaN                                    2.0   
2                       NaN                                    5.0   
3                       NaN                                    NaN   
4                       4.0                                    1.0   

   593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                    4.0                     4.0   
1                                    4.0                     2.0   
2                                    2.0                     NaN   
3                                    NaN                     1.0   
4                                    4.0                     1.0   

   1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
0                  2.0                               2.0  ...   
1                  1.0                               NaN  ...   
2                  4.0                               NaN  ...   
3                  2.0                               3.0  ...   
4                  NaN                               4.0  ...   

   2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
0                               2.0                        NaN   
1                               3.0                        2.0   
2                               NaN                        1.0   
3                               NaN                        4.0   
4                               5.0                        1.0   

   780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
0                                 5.0                       2.0   
1                                 2.0                       NaN   
2                                 4.0                       4.0   
3                                 1.0                       3.0   
4                                 NaN                       4.0   

   1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
0                         NaN                       4.0   
1                         2.0                       NaN   
2                         1.0                       1.0   
3                         5.0                       NaN   
4                         NaN                       3.0   

   527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
0                           2.0                            5.0   
1                           5.0                            1.0   
2                           2.0                            3.0   
3                           5.0                            1.0   
4                           5.0                            5.0   

   1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
0                                   NaN              NaN  
1                                   3.0              NaN  
2                                   1.0              3.0  
3                                   1.0              2.0  
4                                   NaN              NaN  

[5 rows x 22 columns]

In [6]:
# Create a subset data excluding User and Gender columns
mydata_movie = mydata.drop(columns = ['User', 'Gender (1 =F, 0=M)'])

**Question 1** - Mean Rating: Calculate the mean rating for each movie, order with the highest rating listed first, and submit the top three (along with the mean scores for the top two).

In [7]:
mydata_movie.mean().sort_values(ascending = False)[0:3]

318: Shawshank Redemption, The (1994)             3.600000
260: Star Wars: Episode IV - A New Hope (1977)    3.266667
541: Blade Runner (1982)                          3.222222
dtype: float64

**Question 2** - Rating Count (popularity): Count the number of ratings for each movie, order with the most number of ratings first, and submit the top three (along with the counts for the top two). 

In [8]:
mydata_movie.count().sort_values(ascending = False)[0:3]

1: Toy Story (1995)                               17
593: Silence of the Lambs, The (1991)             16
260: Star Wars: Episode IV - A New Hope (1977)    15
dtype: int64

**Question 3** - % of ratings 4+ (liking): Calculate the percentage of ratings for each movie that are 4 or higher. Order with the highest percentage first, and submit the top three (along with the percentage for the top two).  Notice that the three different measures of "best" reflect different priorities and give different results; this should help you see why you need to be thoughtful about what metrics you use. 

In [9]:
(mydata_movie[mydata_movie >= 4.0].count()/mydata_movie.count()).sort_values(ascending = False)[0:3]

318: Shawshank Redemption, The (1994)             0.700000
260: Star Wars: Episode IV - A New Hope (1977)    0.533333
3578: Gladiator (2000)                            0.500000
dtype: float64

**Question 4** - Top movies for someone who has seen Toy Story: Calculate movies that most often occur with Movie #1:  Toy Story, using the (x+y)/x method described in class. In other words, for each movie, calculate the percentage of Toy Story raters who also rated that movie. Order with the highest percentage first, and submit the top 3 (along with the correlations for the top two).  Note, you will have ties -- to break the ties, use the lowest-numbered movie as the higher-ranked one.  In other words, if Movies 541 and 318 are tied, then 318 gets the higher rank.

In [10]:
# Delete the 1st row since it is Toy Story
# Movie 593 ranks 2nd and Movie 1210 ranks 3rd as described in the questions
(mydata_movie[mydata_movie['1: Toy Story (1995)'].notnull()].count()/mydata_movie[mydata_movie['1: Toy Story (1995)'].notnull()].shape[0]).sort_values(ascending = False)[1:5]

260: Star Wars: Episode IV - A New Hope (1977)             0.823529
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.764706
593: Silence of the Lambs, The (1991)                      0.764706
780: Independence Day (ID4) (1996)                         0.764706
dtype: float64

**Question 5** - Correlation with Toy Story:  Calculate the correlation between the vectors of ratings for Toy Story and each other movie.  You can use the built-in CORREL() function.  Order by the highest correlation (positive only) and submit the top 3 along with the correlation values for the top 2.  Notice the differences between co-occurrence and correlation; these metrics are showing different types of relationships.

In [11]:
# Delete the 1st row since it is Toy Story
mydata_movie.corr()['1: Toy Story (1995)'].sort_values(ascending = False)[1:4]

318: Shawshank Redemption, The (1994)    0.888523
34: Babe (1995)                          0.811107
296: Pulp Fiction (1994)                 0.709842
Name: 1: Toy Story (1995), dtype: float64

**Question 6** - Mean rating difference by gender:  First, recompute the mean rating for each movie separately for males and for females.  And calculate the overall mean rating (across all ratings) for males and females. Submit the two movies that have the greatest differences (one where men are most above women, and one where women are most above men) along with the differences in average.  Also submit the difference in overall rating averages (female average - male average).

In [12]:
# Top movie that men are most above women
mydata.drop(columns = ['User']).groupby(['Gender (1 =F, 0=M)']).mean().diff().iloc[1,:].sort_values(ascending = False)[-1:]

1198: Raiders of the Lost Ark (1981)   -1.666667
Name: 1, dtype: float64

In [13]:
# Top movie that women are most above men
mydata.drop(columns = ['User']).groupby(['Gender (1 =F, 0=M)']).mean().diff().iloc[1,:].sort_values(ascending = False)[:1]

2396: Shakespeare in Love (1998)    2.107143
Name: 1, dtype: float64

In [30]:
# Difference = women avg score - men avg score
(mydata.drop(columns = ['User']).groupby(['Gender (1 =F, 0=M)']).sum().sum(axis=1)/mydata.drop(columns = ['User']).groupby(['Gender (1 =F, 0=M)']).count().sum(axis=1)).diff()[1]

0.0418566100290092

**Question 7** - Next, compute the % of ratings 4+ separately for males and females.  You'll be asked to submit two movies as above (largest difference in each direction). And again you'll indicate whether men or women are more likely to rate movies 4 stars or above.

In [14]:
mydata_men = mydata[mydata['Gender (1 =F, 0=M)'] == 0].drop(columns = ['User', 'Gender (1 =F, 0=M)'])
mydata_women = mydata[mydata['Gender (1 =F, 0=M)'] == 1].drop(columns = ['User', 'Gender (1 =F, 0=M)'])

In [15]:
# Top movie that men are most above women
((mydata_men[mydata_men >= 4.0].count()/mydata_men.count()) - (mydata_women[mydata_women >= 4.0].count()/mydata_women.count())).sort_values(ascending = False)[:1]

1198: Raiders of the Lost Ark (1981)    0.5
dtype: float64

In [16]:
# Top movie that women are most above men
((mydata_men[mydata_men >= 4.0].count()/mydata_men.count()) - (mydata_women[mydata_women >= 4.0].count()/mydata_women.count())).sort_values(ascending = False)[-1:]

2396: Shakespeare in Love (1998)   -0.75
dtype: float64

In [37]:
# Difference = women 4+ % - men 4+ %
mydata_women[mydata_women >= 4.0].count().sum()/mydata_women.count().sum() - mydata_men[mydata_men >= 4.0].count().sum()/mydata_men.count().sum()

0.082469954413593